File: primes.go
// Author: Raciel Antela Pardo
// Date: 1-11-2023
// Note: This is the initial version I programmed before the assignment requirement was updated. I kept this since it
// actually resulted in the best runtime (500ms) on a Ryzen 5800x compared to my C++ solution (700ms). My guess is that
// the Go compiler is able to optimize this code better along with a more lightweight sync/threading library.

package main

import (
	"fmt"
	"log"
	"math"
	"os"
	"sync"
	"time"
)

const MAX_NUM = 100000000 // Max number to check for primes
const MAX_THREADS = 8     // Max threads to spawn

// Writes the following to a primes.txt file:
// <execution time>  <total number of primes found>  <sum of all primes found>
// <top ten maximum primes, listed in order from lowest to highest>
func writeToFile(primes []bool, durationMs int64) {
	f, err := os.Create("primes.txt")
	if err != nil {
		log.Fatal(err)
	}

	topPrimes := make([]int, 0, 10)
	primesSum := 0
	primesCount := 0

	for i := MAX_NUM; i >= 2; i-- {
		if primes[i] == true {
			if len(topPrimes) < 10 {
				topPrimes = append(topPrimes, i)
			}
			primesSum += i
			primesCount++
		}
	}

	f.WriteString(fmt.Sprintf("%dms %d %d\n", durationMs, primesCount, primesSum))
	for i := 9; i >= 0; i-- {
		f.WriteString(fmt.Sprintf("%d ", topPrimes[i]))
	}
	f.WriteString("\n")
	defer f.Close()
}

// Sets to false all composite numbers in the primes array
func markComposite(primes []bool, start int, end int, wg *sync.WaitGroup) {
	defer wg.Done()
	for i := start; i <= end; i++ {
		if primes[i] == true {
			for j := i * 2; j <= MAX_NUM; j += i {
				primes[j] = false
			}
		}
	}
}

func main() {
	primes := make([]bool, MAX_NUM+1)
	sqrtMaxNum := int(math.Sqrt(MAX_NUM))

	for i := 0; i < MAX_NUM+1; i++ {
		primes[i] = true
	}

	// Spawn 8 threads to mark composite numbers
	start := time.Now()
	var wg sync.WaitGroup
	wg.Add(MAX_THREADS)
	for i := 0; i < MAX_THREADS; i++ {
		start := 2 + i*sqrtMaxNum/MAX_THREADS
		end := start + sqrtMaxNum/MAX_THREADS
		go markComposite(primes, start, end, &wg)
	}
	wg.Wait()

	writeToFile(primes, time.Since(start).Milliseconds())

	fmt.Println("Finished")
}

File: primes.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <chrono>
#include <thread>
#include <mutex>

using namespace std;

const int MAX_NUM = 100000000;
const int MAX_THREADS = 8; // Total threads used
const int MAIN_THREADS = MAX_THREADS * 2 / 3; // Thread count used for most of the work

// Writes the following to a primes.txt file:
// <execution time>  <total number of primes found>  <sum of all primes found>
// <top ten maximum primes, listed in order from lowest to highest>
void writeToFile(vector<int> primes, int64_t durationMs) {
    ofstream f;
    f.open("primes.txt");
    vector<int> topPrimes;
    uint64_t primesSum = 0;
    uint64_t primesCount = 0;

    for (int i = MAX_NUM; i >= 2; i--) {
        if (primes[i]) {
            if (topPrimes.size() < 10) {
                topPrimes.push_back(i);
            }
            primesSum += i;
            primesCount++;
        }
    }

    f << durationMs << "ms " << primesCount << " " << primesSum << endl;
    for (int i = 9; i >= 0; i--) {
        f << topPrimes[i] << " ";
    }
    f << endl;
    f.close();
}

// Sets to false all composite numbers in the primes array
void markComposite(vector<int> &primes, uint64_t start, uint64_t end) {
    auto startTimer = chrono::high_resolution_clock::now();
    for (uint64_t i = start; i <= end; i++) {
        if (primes[i]) {
            for (uint64_t j = i * 2; j <= MAX_NUM; j += i) {
                primes[j] = 0;
            }
        }
    }
    cout << "Start: " << start << " End: " << end << " Time: "
         << chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - startTimer).count()
         << "ms" << endl;
}

int main() {
    vector<int> primes(MAX_NUM + 1, 1);
    auto sqrtMaxNum = uint64_t(sqrt(MAX_NUM));

    // Spawn 8 threads to mark composite numbers
    auto startTimer = chrono::high_resolution_clock::now();
    vector<thread> threads;
    uint64_t prevChunkEnd = 2;
    for (int i = 0; i < MAX_THREADS; i++) {
        // Assign a larger thread count to the smaller multiples (as there are more of them) for better performance
        if (i < MAIN_THREADS) {
            auto end = prevChunkEnd * 2 < sqrtMaxNum ? prevChunkEnd * 2 : sqrtMaxNum;

            threads.emplace_back(markComposite, ref(primes), prevChunkEnd, end);
            prevChunkEnd = end;
        } else {
            auto chunkSize = (sqrtMaxNum - prevChunkEnd) / (MAX_THREADS - MAIN_THREADS);
            auto start = prevChunkEnd + (i - MAIN_THREADS) * chunkSize;
            auto end = start + chunkSize;

            threads.emplace_back(markComposite, ref(primes), start, end);
        }
    }
    for (auto &t: threads) {
        t.join();
    }
    auto endTimer = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(endTimer - startTimer);

    writeToFile(primes, duration.count());

    cout << "Finished in " << duration.count() << "ms" << endl;
}
